src/librustc/session/code_stats.rs

```diff
@@ -0,0 +1,169 @@
+            variants: variants,
+        };
+        self.type_sizes.insert(info);
+    }
+
+    pub fn print_type_sizes(&self) {
+        let mut sorted: Vec<_> = self.type_sizes.iter().collect();
+
+        // Primary sort: large-to-small.
+        // Secondary sort: description (dictionary order)
+        sorted.sort_by(|info1, info2| {
+            // (reversing cmp order to get large-to-small ordering)
+            match info2.overall_size.cmp(&info1.overall_size) {
+                Ordering::Equal => info1.type_description.cmp(&info2.type_description),
+                other => other,
+            }
+        });
+
+        for info in &sorted {
+            println!("print-type-size type: `{}`: {} bytes, alignment: {} bytes",
+                     info.type_description, info.overall_size, info.align);
+            let indent = "    ";
+
+            let discr_size = if let Some(discr_size) = info.opt_discr_size {
+                println!("print-type-size {}discriminant: {} bytes",
+                         indent, discr_size);
+                discr_size
+            } else {
+                0
+            };
+
+            // We start this at discr_size (rather than 0) because
+            // things like C-enums do not have variants but we still
+            // want the max_variant_size at the end of the loop below
+            // to reflect the presence of the discriminant.
+            let mut max_variant_size = discr_size;
+
+            let struct_like = match info.kind {
+                DataTypeKind::Struct | DataTypeKind::Closure => true,
+                DataTypeKind::Enum | DataTypeKind::Union => false,
+            };
+            for (i, variant_info) in info.variants.iter().enumerate() {
+                let VariantInfo { ref name, kind: _, align: _, size, ref fields } = *variant_info;
+                let indent = if !struct_like {
+                    let name = match name.as_ref() {
+                        Some(name) => format!("{}", name),
+                        None => format!("{}", i),
+                    };
+                    println!("print-type-size {}variant `{}`: {} bytes",
+                             indent, name, size - discr_size);
+                    "        "
+                } else {
+                    assert!(i < 1);
+                    "    "
+                };
+                max_variant_size = cmp::max(max_variant_size, size);
+
+                let mut min_offset = discr_size;
+                for field in fields {
+                    let FieldInfo { ref name, offset: _lying_offset, size, align } = *field;
+                    let offset = roundup(min_offset, align);
+                    if offset != _lying_offset {
+                        // Experiments indicate that the offset extracted from Layout does
+                        // not always correctly account for padding. Thus the name :(
+                        debug!("offset: {} _lying_offset: {}", offset, _lying_offset);
```

Comment:
It reproduces regularly on the test/ui/print_type_sizes.rs test. Output is like this when you turn the debug logging on:

```
print-type-size type: `E1`: 24 bytes, alignment: 8 bytes
print-type-size     discriminant: 4 bytes
print-type-size     variant `A`: 16 bytes
DEBUG:rustc::session::code_stats: offset: 8 _lying_offset: 0
print-type-size         padding bytes: 4
print-type-size         field `.0`: 8 bytes, alignment: 8 bytes
DEBUG:rustc::session::code_stats: offset: 16 _lying_offset: 8
print-type-size         field `.1`: 4 bytes
print-type-size     variant `B`: 8 bytes
DEBUG:rustc::session::code_stats: offset: 4 _lying_offset: 0
print-type-size         field `.0`: 8 bytes
print-type-size     end padding bytes: 4
print-type-size type: `E2`: 16 bytes, alignment: 8 bytes
print-type-size     discriminant: 4 bytes
print-type-size     variant `A`: 12 bytes
DEBUG:rustc::session::code_stats: offset: 4 _lying_offset: 0
print-type-size         field `.0`: 4 bytes
DEBUG:rustc::session::code_stats: offset: 8 _lying_offset: 4
print-type-size         field `.1`: 8 bytes
print-type-size     variant `B`: 8 bytes
DEBUG:rustc::session::code_stats: offset: 4 _lying_offset: 0
print-type-size         field `.0`: 8 bytes
print-type-size type: `S`: 8 bytes, alignment: 4 bytes
print-type-size     field `.a`: 1 bytes
print-type-size     field `.b`: 1 bytes
print-type-size     padding bytes: 2
print-type-size     field `.g`: 4 bytes, alignment: 4 bytes
```

(My assertion that the reported offset is "lying" is based on manual inspection of a value that I transmuted to a tuple.)

I've been assuming that this is not currently as serious a bug as one might think because I've been assuming that nothing important is *using* the offset values computed by the layout module. However, as I mentioned to @nnethercote in an (outdated) comment above,

> Whenever we get around to rearranging the field order, we will have to fix the layout module to actually compute/maintain correct offset values, and we can revise this accordingly then.